app/src/main/java/com/google/android/samples/socialite/ui/home/timeline/TimelineViewModel.kt

```diff
@@ -105,31 +121,59 @@ class TimelineViewModel @Inject constructor(
                 it.repeatMode = ExoPlayer.REPEAT_MODE_ONE
                 it.playWhenReady = true
                 it.addListener(videoSizeListener)
+                it.addListener(firstFrameListener)
             }
 
         videoRatio = null
         player = newPlayer
+
+        if (enablePreloadManager) {
+            initPreloadManager(loadControl)
+        }
+    }
+
+    private fun initPreloadManager(loadControl: DefaultLoadControl) {
+        preloadManager =
+            PreloadManagerWrapper.build(
+                (player as ExoPlayer).applicationLooper,
+                loadControl,
+                application.applicationContext,
+            )
+        preloadManager.setPreloadWindowSize(5)
+
+        // Add videos to preload
+        if (media.isNotEmpty()) {
+            preloadManager.init(media)
+        }
     }
 
     fun releasePlayer() {
         player?.apply {
             removeListener(videoSizeListener)
             release()
         }
-
+        if (enablePreloadManager) {
+            preloadManager.release()
+        }
         videoRatio = null
         player = null
     }
 
-    fun changePlayerItem(uri: Uri?) {
+    fun changePlayerItem(uri: Uri?, currentPlayingIndex: Int) {
         if (player == null) return
 
         player?.apply {
             stop()
             videoRatio = null
             if (uri != null) {
                 setMediaItem(MediaItem.fromUri(uri))
```

Comment:
`Player.setMediaItem` will have the player internally to create a new `MediaSource` and play it, but this is not the one `PreloadManager` has preloaded. I think here you need to call:

```
setMediaSource(preloadManager.getMediaSource(MediaItem.fromUri(uri)))
```

Reply:
The changePlayerItem method does check if preload flag is enabled (enablePreloadManager) and if the media source is preloaded, only in that case the, we get the MediaSource from the preload manager. In other cases, we get the media source directly.

Reply:
I think Tianyi might have the right recommendation here. Perhaps we need to change the order of operations to be something like:

```
if (enablePreloadManager) {
  // Configure PreloadManager
  preloadManager.setCurrentPlayingIndex(currentPlayingIndex)
  preloadManager.addMediaItem(MediaItem.fromUri(uri), currentPlayingIndex)
  // Set Player to play media item from PreloadManager
  setMediaSource(preloadManager.getMediaSource(MediaItem.fromUri(uri)))
} else {
  // Play media item directly
  setMediaItem(MediaItem.fromUri(uri))
}
prepare()
```

WDYT?